git/git.go

```diff
@@ -0,0 +1,183 @@
+// for a "tag-1" tag.
+func (c *Commit) String() string {
+	if short := strings.SplitAfterN(c.Reference, "/", 3); len(short) == 3 {
+		return fmt.Sprintf("%s/%s", short[2], c.Hash)
+	}
+	return fmt.Sprintf("HEAD/%s", c.Hash)
+}
+
+// Verify the Signature of the commit with the given key rings.
+// It returns the fingerprint of the key the signature was verified
+// with, or an error.
+func (c *Commit) Verify(keyRing ...string) (string, error) {
+	if c.Signature == "" {
+		return "", fmt.Errorf("commit does not have a PGP signature")
+	}
+
+	for _, r := range keyRing {
+		reader := strings.NewReader(r)
+		keyring, err := openpgp.ReadArmoredKeyRing(reader)
+		if err != nil {
+			return "", fmt.Errorf("unable to read armored key ring: %w", err)
+		}
+		signer, err := openpgp.CheckArmoredDetachedSignature(keyring, bytes.NewBuffer(c.Encoded), bytes.NewBufferString(c.Signature), nil)
+		if err == nil && len(signer.PrimaryKey.Fingerprint) >= 20 {
+			return fmt.Sprintf("%X", signer.PrimaryKey.Fingerprint[12:20]), nil
+		}
+	}
+	return "", fmt.Errorf("unable to verify commit with any of the given key rings")
+}
+
+// ShortMessage returns the first 50 characters of a commit subject.
+func (c *Commit) ShortMessage() string {
+	subject := strings.Split(c.Message, "\n")[0]
+	r := []rune(subject)
+	if len(r) > 50 {
+		return fmt.Sprintf("%s...", string(r[0:50]))
+	}
+	return subject
+}
+
+// RepositoryReader knows how to perform local and remote read operations
+// to a Git repository.
+type RepositoryReader interface {
+	// Clone clones a repository from the provided url using the options provided.
+	Clone(ctx context.Context, url string, checkoutOpts CheckoutOptions) (*Commit, error)
+	// IsClean returns whether the working tree is clean.
+	IsClean() (bool, error)
+	// Head returns the hash of the current HEAD of the repo.
+	Head() (string, error)
+	// Path returns the path of the repository.
+	Path() string
+	// Cleanup frees any resources that need to be freed at the end of
+	// the client's lifecycle.
+	Cleanup()
+}
+
+// RepositoryWriter knows how to perform local and remote write operations
+// to a Git repository.
+type RepositoryWriter interface {
+	// Init initializes a repository at the configured path with the origin
+	// remote set to url on the provided branch.
+	Init(ctx context.Context, url, branch string) error
+	// Push pushes the current branch of the repository to origin.
+	Push(ctx context.Context) error
+	// SwitchBranch switches the active branch of the repository to the
+	// provided branch. If the branch doesn't exist, it is created.
+	SwitchBranch(ctx context.Context, branch string) error
+	// WriteFile creates a new file at the given path, reads from the reader and
+	// writes the content to the file. If the file already exists, it's contents
+	// are overwritten.
+	WriteFile(path string, reader io.Reader) error
+	// Commit commits any changes made to the repository.
+	Commit(info Commit, signer *openpgp.Entity) (string, error)
```

Comment:
Shouldn't signer be an optional parameter?
Maybe we can have options for some other methods above as well. For things that we assume defaults, like "origin" being the default remote when pushing, "main" being the default branch when initializing, bare repo option etc.

Reply:
`signer` _is_ an optional parameter, I'll update the method docs to reflect this.

Reply:
You can define the default branch of the repo through `Init(url, branch)`. I don't see why we need to support bare repos as bare repos are primarily meant for git servers and not for users.